subprojects/core/src/main/java/org/gradle/execution/SelectedTaskExecutionAction.java

```diff
@@ -38,8 +39,9 @@ public ExecutionResult<Void> execute(GradleInternal gradle, FinalizedExecutionPl
     private void bindAllReferencesOfProject(TaskExecutionGraph graph) {
         Set<Project> seen = Sets.newHashSet();
         for (Task task : graph.getAllTasks()) {
-            if (seen.add(task.getProject())) {
-                ProjectInternal projectInternal = (ProjectInternal) task.getProject();
+            Project taskProject = ((TaskInternal) task).getProjectUnchecked();
+            if (seen.add(taskProject)) {
+                ProjectInternal projectInternal = (ProjectInternal) taskProject;
```

Comment:
`getProjectUnchecked` already returns a `ProjectInternal`. No need to cast.

Reply:
Done in https://github.com/gradle/gradle/pull/23210/commits/7c7db9ff2439cff8a895961e4fc2d26bfe2142a7